File: leetcode 1776.py
from typing import List

class Solution:
    def getCollisionTimes(self, cars: List[List[int]]) -> List[float]:
        n = len(cars)
        stk = []
        ans = [-1.0] * n
        for i in range(n - 1, -1, -1):
            pi = cars[i][0]; vi = cars[i][1]
            while stk:
                if stk and cars[stk[-1]][1] >= vi:
                    stk.pop()
                    continue
                # 没车了，说明i号车追不上前面的了
                if not stk:
                    break
                # 有车，但是需要变成追击对象
                T = (cars[stk[-1]][0] - pi) / (vi - cars[stk[-1]][1])
                if ans[stk[-1]] > 0 and T > ans[stk[-1]]:
                    stk.pop()
                    continue
                ans[i] = T
                break
            stk.append(i)
        return ans

File: leetcode 1776.cpp
class Solution {
public:
  vector<double> getCollisionTimes(vector<vector<int>> &cars) {
    int n = cars.size();
    std::vector<int> stk;
    std::vector<double> ans(n, -1);
    for (int i = n - 1; ~i; i--) {
      int pi = cars[i][0], vi = cars[i][1];
      while (stk.size()) {
        if (cars[stk.back()][1] >= vi) {
          stk.pop_back();
          continue;
        }
        if (stk.empty())
          break;
        // 当前车能追上栈顶的车了
        double T =
            (cars[stk.back()][0] - pi) * 1.0 / (vi - cars[stk.back()][1]);
        if (ans[stk.back()] > 0 && T > ans[stk.back()]) {
          stk.pop_back();
          continue; // 栈顶的车在被追上之前追上了更前面的车，因此需要用更前面的车为基准来计算追击问题
        }
        ans[i] = T;
        break;
      }
      stk.push_back(i);
    }
    return ans;
  }
};